diff --git a/st.c b/st.c
index 2335fdd..a5995fb 100644
--- a/st.c
+++ b/st.c
@@ -32,10 +32,10 @@ typedef struct st_packed_entry {
 
 #define STATIC_ASSERT(name, expr) typedef int static_assert_##name##_check[(expr) ? 1 : -1];
 
-#define ST_DEFAULT_MAX_DENSITY 5
+#define ST_DEFAULT_MAX_DENSITY 2
 #define ST_DEFAULT_INIT_TABLE_SIZE 11
 #define ST_DEFAULT_SECOND_TABLE_SIZE 19
-#define ST_DEFAULT_PACKED_TABLE_SIZE 18
+#define ST_DEFAULT_PACKED_TABLE_SIZE 19
 #define PACKED_UNIT (int)(sizeof(st_packed_entry) / sizeof(st_table_entry*))
 #define MAX_PACKED_HASH (int)(ST_DEFAULT_PACKED_TABLE_SIZE * sizeof(st_table_entry*) / sizeof(st_packed_entry))
 
