diff --git a/test/etc/test_etc.rb b/test/etc/test_etc.rb
index c105122..44b6676 100644
--- a/test/etc/test_etc.rb
+++ b/test/etc/test_etc.rb
@@ -76,6 +76,7 @@ class TestEtc < Test::Unit::TestCase
   end
 
   def test_getgrgid
+    # skip "Broken Etc.getgrgid on this machine" if ENV['RUBY_TEST_OPTIONS_BROKEN_GETGRGID']=='1'
     # group database is not unique on GID, and which entry will be
     # returned by getgrgid() is not specified.
     groups = Hash.new {[]}
@@ -92,6 +93,7 @@ class TestEtc < Test::Unit::TestCase
   end
 
   def test_getgrnam
+    skip "Broken Etc.getgrnam on this machine" if ENV['RUBY_TEST_OPTIONS_BROKEN_GETGRNAM']=='1'
     groups = {}
     Etc.group do |s|
       groups[s.name] ||= s unless /\A\+/ =~ s.name
diff --git a/test/ruby/test_process.rb b/test/ruby/test_process.rb
index c77cbab..0c9d537 100644
--- a/test/ruby/test_process.rb
+++ b/test/ruby/test_process.rb
@@ -1570,6 +1570,7 @@ class TestProcess < Test::Unit::TestCase
 
   def test_execopts_gid
     skip "Process.groups not implemented on Windows platform" if windows?
+    skip "Broken Etc.getgrgid on this machine" if ENV['RUBY_TEST_OPTIONS_BROKEN_GETGRGID']=='1'
     feature6975 = '[ruby-core:47414]'
 
     [30000, *Process.groups.map {|g| g = Etc.getgrgid(g); [g.name, g.gid]}].each do |group, gid|
